groups/etcd_group_service.go

```diff
@@ -0,0 +1,284 @@
+package groups
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/namespace"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/constants"
+	"github.com/topfreegames/pitaya/logger"
+)
+
+var (
+	clientInstance     *clientv3.Client
+	transactionTimeout time.Duration
+	etcdOnce           sync.Once
+)
+
+// EtcdGroupService base ETCD struct solution
+type EtcdGroupService struct {
+}
+
+// NewEtcdGroupService returns a new group instance
+func NewEtcdGroupService(conf *config.Config, clientOrNil *clientv3.Client) (*EtcdGroupService, error) {
+	err := initClientInstance(conf, clientOrNil)
+	if err != nil {
+		return nil, err
+	}
+	return &EtcdGroupService{}, err
+}
+
+func initClientInstance(config *config.Config, clientOrNil *clientv3.Client) error {
+	var err error
+	etcdOnce.Do(func() {
+		if clientOrNil != nil {
+			clientInstance = clientOrNil
+		} else {
+			clientInstance, err = createBaseClient(config)
+		}
+		if err != nil {
+			logger.Log.Fatalf("error initializing singleton etcd client in groups: %s", err.Error())
+			return
+		}
+		transactionTimeout = config.GetDuration("pitaya.groups.etcd.transactiontimeout")
+	})
+	return err
+}
+
+func createBaseClient(config *config.Config) (*clientv3.Client, error) {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   config.GetStringSlice("pitaya.groups.etcd.endpoints"),
+		DialTimeout: config.GetDuration("pitaya.groups.etcd.dialtimeout"),
+	})
+	if err != nil {
+		return nil, err
+	}
+	cli.KV = namespace.NewKV(cli.KV, config.GetString("pitaya.groups.etcd.prefix"))
+	return cli, nil
+}
+
+func groupKey(groupName string) string {
+	return fmt.Sprintf("groups/%s", groupName)
+}
+
+func memberKey(groupName, uid string) string {
+	return fmt.Sprintf("%s/uids/%s", groupKey(groupName), uid)
+}
+
+func getGroupKV(ctx context.Context, groupName string) (*mvccpb.KeyValue, error) {
+	ctxT, cancel := context.WithTimeout(ctx, transactionTimeout)
+	etcdRes, err := clientInstance.Get(ctxT, groupKey(groupName))
+	cancel()
+	if err != nil {
+		return nil, err
+	}
+	if etcdRes.Count == 0 {
+		return nil, constants.ErrGroupNotFound
+	}
+	return etcdRes.Kvs[0], nil
+}
+
+func (c *EtcdGroupService) createGroup(ctx context.Context, groupName string, leaseID clientv3.LeaseID) error {
+	var etcdRes *clientv3.TxnResponse
+	var err error
+
+	ctxT, cancel := context.WithTimeout(ctx, transactionTimeout)
+	if leaseID != 0 {
+		etcdRes, err = clientInstance.Txn(ctxT).
+			If(clientv3.Compare(clientv3.CreateRevision(groupKey(groupName)), "=", 0)).
+			Then(clientv3.OpPut(groupKey(groupName), "", clientv3.WithLease(leaseID))).
+			Commit()
+	} else {
+		etcdRes, err = clientInstance.Txn(ctxT).
+			If(clientv3.Compare(clientv3.CreateRevision(groupKey(groupName)), "=", 0)).
+			Then(clientv3.OpPut(groupKey(groupName), "")).
+			Commit()
+	}
+	cancel()
```

Comment:
This is an interesting question. If I use defer, it will only be called at the end of the whole function which it has been called and I only want the transaction operation to be with timeout. That's why I inserted it right after the method call, to be certain that it was called exactly after. Also, the doc says that the timeout should be called asap. Should I use defer?